src/dtls.c

```diff
@@ -290,60 +343,385 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+            if (tlsx.size != 0) {
+                /* Let's just assume that the binders are correct here. We will
+                 * actually verify this in the stateful part of the processing
+                 * and if they don't match we will error out there anyway. */
+                byte modes;
+
+                ret = TlsxFindByType(&tlsx, TLSX_PSK_KEY_EXCHANGE_MODES,
+                                     ch->extension);
+                if (ret != 0)
+                    goto dtls13_cleanup;
+                if (tlsx.size == 0)
+                    ERROR_OUT(MISSING_HANDSHAKE_DATA, dtls13_cleanup);
+                ret = TLSX_PskKeyModes_Parse_Modes(tlsx.elements, tlsx.size,
+                                                  client_hello, &modes);
+                if (ret != 0)
+                    goto dtls13_cleanup;
+                if ((modes & (1 << PSK_DHE_KE)) && !ssl->options.noPskDheKe) {
+                    if (!haveKS)
+                        ERROR_OUT(MISSING_HANDSHAKE_DATA, dtls13_cleanup);
+                    doKE = 1;
+                }
+                else {
+                    if ((modes & (1 << PSK_KE)) == 0)
+                        ERROR_OUT(PSK_KEY_ERROR, dtls13_cleanup);
+                }
+            }
+#endif
+
+#if defined(HAVE_SESSION_TICKET) || !defined(NO_PSK)
+            if (usePSK) {
+                if (pskInfo->isValid) {
+                    cs.cipherSuite0 = pskInfo->cipherSuite0;
+                    cs.cipherSuite = pskInfo->cipherSuite;
+                }
+                else {
+                    /* Only support the default ciphersuite for PSK */
+                    cs.cipherSuite0 = TLS13_BYTE;
+                    cs.cipherSuite = WOLFSSL_DEF_PSK_CIPHER;
+                }
+
+                if (doKE) {
+                    byte searched = 0;
+                    ret = TLSX_KeyShare_Choose(ssl, parsedExts, &cs.clientKSE,
+                            &searched);
+                    if (ret != 0)
+                        goto dtls13_cleanup;
+                    if (cs.clientKSE == NULL && searched)
+                        cs.doHelloRetry = 1;
+                }
+            }
+            else
+#endif
+            {
+                ret = MatchSuite_ex(ssl, &suites, &cs, parsedExts);
+                if (ret < 0) {
+                    WOLFSSL_MSG("Unsupported cipher suite, ClientHello");
+                    SendAlert((WOLFSSL*)ssl, alert_fatal, handshake_failure);
+                    goto dtls13_cleanup;
+                }
+            }
+            if (cs.doHelloRetry) {
+                ret = TLSX_KeyShare_SetSupported(ssl, &parsedExts);
+                if (ret != 0)
+                    goto dtls13_cleanup;
+            }
+
+            /* Need to remove the keyshare ext if we are not doing PSK and we
+             * found a common group. */
+            if (cs.clientKSE != NULL
+#if defined(HAVE_SESSION_TICKET) || !defined(NO_PSK)
+                    && !usePSK
```

Comment:
In HRR we can't send keyshare when we don't want to negotiate the curve. Only when we have to negotiate we should send the keyshare.

Reply:
but if using psk with dhe we still want to negotiate the curve

Reply:
Yes but we do that later in the stateful CH. I don't think we can negotiate that here.

Reply:
In the stateful CH it will be too late. We have to ask the client for a good KS value now as we can't send two HelloRetryRequest in the same connection.

Reply:
This has been refactored to depend on a `cs.doHelloRetry` variable. Take another look if it makes sense.